File: backend/runtime/actions/send_email.py
"""Send email action dispatcher."""

import logging

from runtime.actions.base import ActionDispatcher

logger = logging.getLogger(__name__)


class SendEmailDispatcher(ActionDispatcher):
    """Resolves email fields and logs the dispatch (actual sending is integration-dependent)."""

    async def execute(self, config: dict) -> dict:
        to = self.resolver.resolve_string(config.get("to", ""))
        subject = self.resolver.resolve_string(config.get("subject", ""))
        template = self.resolver.resolve_string(config.get("template", ""))
        body = self.resolver.resolve_string(config.get("body", ""))

        logger.info("Email action — to: %s, subject: %s", to, subject)

        return {
            "action_type": "send_email",
            "to": to,
            "subject": subject,
            "template": template,
            "body": body,
            "result": "sent",
        }

File: backend/templates/workflow-engine/actions/send-email.ts
/** Send email action dispatcher. */

import { ActionDispatcher } from './base';

export class SendEmailDispatcher extends ActionDispatcher {
  async execute(config: Record<string, unknown>): Promise<Record<string, unknown>> {
    const to = this.resolver.resolveString((config.to as string) ?? '');
    const subject = this.resolver.resolveString((config.subject as string) ?? '');
    const template = this.resolver.resolveString((config.template as string) ?? '');
    const body = this.resolver.resolveString((config.body as string) ?? '');

    console.log(`[workflow] Email action — to: ${to}, subject: ${subject}`);

    return {
      action_type: 'send_email',
      to,
      subject,
      template,
      body,
      result: 'sent',
    };
  }
}
